operator/src/main/java/com/instaclustr/cassandra/operator/jmx/CassandraConnectionFactory.java

```diff
@@ -0,0 +1,83 @@
+package com.instaclustr.cassandra.operator.jmx;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalNotification;
+import com.google.common.net.InetAddresses;
+import io.kubernetes.client.models.V1Pod;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.management.remote.JMXConnectionNotification;
+import javax.management.remote.JMXConnector;
+import javax.management.remote.JMXConnectorFactory;
+import javax.management.remote.JMXServiceURL;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+public class CassandraConnectionFactory {
+    private static final Logger logger = LoggerFactory.getLogger(CassandraConnectionFactory.class);
+
+    private final LoadingCache<InetAddress, CassandraConnection> connectionCache = CacheBuilder.<InetAddress, CassandraConnection>newBuilder()
+            .expireAfterWrite(10, TimeUnit.MINUTES)
+            .removalListener((RemovalNotification<InetAddress, CassandraConnection> notification) -> {
+                try {
+                    notification.getValue().close();
+
+                } catch (final IOException e) {
+                    logger.warn("Failed to close JMX connection to {}.", notification.getKey(), e);
+                }
+            })
+            .build(new CacheLoader<InetAddress, CassandraConnection>() {
+                @Override
+                public CassandraConnection load(@Nonnull final InetAddress key) throws Exception {
+                    final JMXServiceURL serviceURL = new JMXServiceURL(String.format("service:jmx:rmi:///jndi/rmi://%s:7199/jmxrmi", InetAddresses.toAddrString(key)));
+
+                    logger.debug("Establishing JMX connection to {} ({})", key, serviceURL);
+
+                    final JMXConnector jmxConnector = JMXConnectorFactory.connect(serviceURL, null);
+
+                    // observe connections and when they fail/close, invalidate the cache
+                    jmxConnector.addConnectionNotificationListener((notification, handback) -> {
+                        if (!(notification instanceof JMXConnectionNotification)) {
+                            return;
+                        }
+
+                        final JMXConnectionNotification connectionNotification = (JMXConnectionNotification) notification;
+                        switch (connectionNotification.getType()) {
```

Comment:
Cassandra will lose notifications (esp decom and repair) like no tomorrow, we will want to log this, though probably not relevant for the cache

Reply:
True, but this listener is purely for observing the JMX connection closing/failing so that it can be expunged from the cache.

The current design doesn't rely on recieving decommission notifications. The `CassandraHealthCheckService` periodically polls the operation mode (`NORMAL`, `JOINING`,`DECOMM..`, etc) of every node and will broadcast an event if this mode changes. This broadcast allows the `ControllerService` to reconcile when the node switches to `DECOMMISSIONED`.

We could listen for decom notifications as a way to reduce polling (current interval is 1min) and be more "reactive" as the reconcile could occur as soon as the node switches to decommissioned, falling back to polling if we miss notifications.

Reply:
Ok that makes sense, I think polling only is fine and less likely to get in a weird state / less complicated that listening for notifications then falling back.

LGTM.